### PR TITLE
feat: Qdrant instrumentation

### LINKED ISSUE #345: 🚀 Feature: Support qdrant vector database
### Which component is this feature for?

Traceloop SDK

### 🔖 Feature description

Support tracing of calls to the [qdrant](https://github.com/qdrant/qdrant) vector database

### 🎤 Why is this feature needed ?

For users of qdrant

### ✌️ How do you aim to achieve this?

A new qdrant extension

### 🔄️ Additional Information

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Anush008**
Fixes #345 

## Description

This PR adds support for [Qdrant ](https://qdrant.tech/) to be traced by OpenLLMetry by introducing a new instrumentation module `opentelemetry-instrumentation-qdrant`.

Supports `QdrantClient` and `AsyncQdrantClient` from the `qdrant-client >= 1.7` package.

Screenshot 2024-01-26 at 22 01 21

Screenshot 2024-01-26 at 21 58 10

### CODE REVIEW

**Comment by nirga** [comment]
Thanks @Anush008! Can you add tests to the new instrumentation under traceloop-sdk?

**Comment by Anush008** [comment]
Hi @nirga. They are here.
https://github.com/traceloop/openllmetry/blob/5e5f5eb8f933256d3a560f83f67641da5e86ab13/packages/traceloop-sdk/tests/test_qdrant_instrumentation.py

**Comment by Anush008** [comment]
e0e83551022cc5723925f79bca7c2f0858dd4dc7 is breaking all the sdk tests.

**Comment by Anush008** [comment]
e0e83551022cc5723925f79bca7c2f0858dd4dc7 is breaking all the SDK tests.

<details>
<summary>Traceback</summary>
    @pytest.fixture(scope="session")
    def exporter():
        exporter = InMemorySpanExporter()
>       Traceloop.init(
            app_name="test",
            resource_attributes={"something": "yes"},
            disable_batch=True,
            exporter=exporter,
        )

tests/conftest.py:12:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
traceloop/sdk/__init__.py:152: in init
    Traceloop.__tracer_wrapper = TracerWrapper(
traceloop/sdk/tracing/tracing.py:111: in __new__
    init_instrumentations()
traceloop/sdk/tracing/tracing.py:265: in init_instrumentations
    init_haystack_instrumentor()
traceloop/sdk/tracing/tracing.py:345: in init_haystack_instrumentor
    instrumentor.instrument()
.venv/lib/python3.9/site-packages/opentelemetry/instrumentation/instrumentor.py:114: in instrument
    result = self._instrument(  # pylint: disable=assignment-from-no-return
../opentelemetry-instrumentation-haystack/opentelemetry/instrumentation/haystack/__init__.py:71: in _instrument
    wrap_function_wrapper(
.venv/lib/python3.9/site-packages/wrapt/patches.py:114: in wrap_function_wrapper
    return wrap_object(module, name, FunctionWrapper, (wrapper,))
.venv/lib/python3.9/site-packages/wrapt/patches.py:60: in wrap_object
    (parent, attribute, original) = resolve_path(module, name)
.venv/lib/python3.9/site-packages/wrapt/patches.py:17: in resolve_path
    __import__(module)
.venv/lib/python3.9/site-packages/haystack/__init__.py:11: in <module>
    from haystack.nodes.base import BaseComponent
.venv/lib/python3.9/site-packages/haystack/nodes/__init__.py:3: in <module>
    from haystack.nodes.answer_generator import BaseGenerator, OpenAIAnswerGenerator
.venv/lib/python3.9/site-packages/haystack/nodes/answer_generator/__init__.py:2: in <module>
    from haystack.nodes.answer_generator.openai import OpenAIAnswerGenerator
.venv/lib/python3.9/site-packages/haystack/nodes/answer_generator/openai.py:6: in <module>
    from haystack.environment import HAYSTACK_REMOTE_API_TIMEOUT_SEC
.venv/lib/python3.9/site-packages/haystack/environment.py:15: in <module>
    import transformers
.venv/lib/python3.9/site-packages/transformers/__init__.py:26: in <module>
    from . import dependency_versions_check
.venv/lib/python3.9/site-packages/transformers/dependency_versions_check.py:16: in <module>
    from .utils.versions import require_version, require_version_core
.venv/lib/python3.9/site-packages/transformers/utils/__init__.py:18: in <module>
    from huggingface_hub import get_full_repo_name  # for backward compatibility
.venv/lib/python3.9/site-packages/huggingface_hub/__init__.py:370: in __getattr__
    submod = importlib.import_module(submod_path)
../../../../.pyenv/versions/3.9.5/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
.venv/lib/python3.9/site-packages/huggingface_hub/hf_api.py:48: in <module>
    from tqdm.auto import tqdm as base_tqdm
</details>

**Comment by nirga** [comment]
Thanks @Anush008! I'd consider adding instrumentations for the vectors themselves, as we do in #344 for Chroma now.
Reg. the tests - I've triggered their runs now.

**Comment by nirga** [comment]
Also, can you please attach screenshots from Traceloop (or any other observability platform) showing how this instrumentation works?

File: packages/opentelemetry-instrumentation-qdrant/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+[tool.coverage.report]
+exclude_lines = [ "if TYPE_CHECKING:" ]
+show_missing = true
+
+[tool.poetry]
+name = "opentelemetry-instrumentation-qdrant"
+version = "0.10.1"
+description = "OpenTelemetry Qdrant instrumentation"
+authors = [
+  "Gal Kleinman <gal@traceloop.com>",
```
Comment (paolorechia) [inline packages/opentelemetry-instrumentation-qdrant/pyproject.toml:14]: :question: Should contributors that create an instrumentation put themselves as an author here? :thinking:

**Review by paolorechia — COMMENTED** [review]
Overall this code looks great from my perspective.

I find helpful to see the full trace exported to console (https://github.com/traceloop/openllmetry/pull/368#issuecomment-1919532153) - but that's just maybe - I'm also new to this codebase :)

File: packages/opentelemetry-instrumentation-qdrant/pyproject.toml (hunk 1 above)
Comment (nirga) [reply packages/opentelemetry-instrumentation-qdrant/pyproject.toml:14]: I don't have a strong opinion here. In general I'm fine with putting the main author here. As long as the license is APACHE it doesn't affect or limit the usage of the instrumentation.

**Comment by Anush008** [comment]
Hi @nirga. I got on with other tasks after the tests ran. Could I get a review please and I'll proceed from there?

I didn't feel strongly about adding the embeddings to the trace because the request would be bloated with large embedding dimensions and I think people wouldn't really want to see every embedding literal value.

**Comment by nirga** [comment]
@Anush008 overall this looks great. I just have 2 comments:
1. Can you rebase?
2. Can you move the test file to be in the instrumentation package?

**Review by nirga — APPROVED** [review]

**Comment by Anush008** [comment]
🎉🎉🎉

**Pull request merged**